src/arrays/quarray.jl

```diff
@@ -0,0 +1,148 @@
+import Base: size,
+    length,
+    getindex,
+    similar,
+    in,
+    ctranspose,
+    transpose,
+    summary,
+    zeros,
+    eye,
+    #TODO: Implement the below operations
+    *,.*,
+    /,./,
+    +,.+,
+    -,.-,
+    kron
+
+abstract AbstractQuArray{B<:(AbstractBasis...), T, N} <: AbstractArray{T,N}
+
+###########
+# QuArray #                 
+###########
+    # B<:(AbstractBasis...) allows us to have a basis for each dimension, 
+    # and gives us a less ambiguous way to determine if a QuArray will act
+    # like a vector, matrix, or tensor compared to just looking at the 
+    # dimension parameter N. 
+    type QuArray{B<:(AbstractBasis...), T, N, A} <: AbstractQuArray{B, T, N}
+        coeffs::A
+        bases::B
+        function QuArray(coeffs::AbstractArray{T, N}, bases::B) 
+            if checkbases(coeffs, bases) 
+                new(coeffs, bases)
+            else 
+                error("Coefficient array does not conform to input bases")
+            end
+        end
+    end
+
```

Comment:
Nice! I guess I would have asked you about `NTuple`s and `QuArray`s anyways :-) I also think that this change makes the definition of `QuArray`s much cleaner and nicer. The only objection I can think of is your point 1), because it "restricts" us to cartesian tensors only (ie we don't distinguish between raised and lower indices anymore). However, I would say that this is a reasonable compromise, since it will be the most common use case(?).